### ISSUE TITLE
Provide an OCP API for apps to load their initial state

### ISSUE THREAD

**Issue by ChristophWurst** [nextcloud/server#13488]
**Is your feature request related to a problem? Please describe.**
Apps usually have some kind of state for the initial rendering of their UI or for a background task. From what I see in our existing code (server and apps), most dynamic code loads its state asynchronously. The static parts (rendered on the server) usually have their state baked into the resulting HTML, hence no data is fetched.

This causes additional requests. Therefore it takes longer for the page to load, the page does not look ready right away, lots of loading animations are shown (best example: the settings pages) and it just gives the application a bit of an instable/fragile appearance.

**Describe the solution you'd like**
Apps (and the server) should try to load data eagerly with the page to be able to render immediately. While there is a bit of an overhead to load all data at once on the server-side, I am convinced that in total there it's less expensive (in terms of computational power) to load all data once than in individual requests.

It's already possible to do that. The usual trick is to fill a hidden input field with a serialized JSON clob. On the client-side, this clob is deserialized into a JS object. An example of this trick can be seen here: https://github.com/nextcloud/server/pull/13481.

This trick above works for apps that load their own template or have a template rendered somewhere else like the settings. However, it does not work for apps that are active on server or other app's pages because there they have no influence on the rendered template.

I would therefore like to propose a new server-side API to store JSON-serializable data that is then merged into a single serialized clob and proved via the client-side OCP.

**Describe alternatives you've considered**
I don't have any at the moment except for not improving this.

**Additional context**
This helps with the "we have way too many requests on page load" issue.

cc @nextcloud/javascript let me know what you think. I'd work on this ASAP so that we can use this in Nextcloud 16 already and have pages load faster :rocket:

**Comment by rullzer** [comment]
As discussed good idea!

For example notifications could already tell if the initial load is needed etc!

**Comment by skjnldsv** [comment]
This is what we do for the users management settings! A standard would be very nice yes! :)

**Comment by MorrisJobke** [comment]
So this is then how it should work:

On the backend it does this:

```php
$eventDispatcher->addListener(OCP\IConfig::FETCH_CONFIG,
	function(GenericEvent $event) {
		/** @var FetchConfigSubject $subject */
		$subject = $event->getSubject();
		$subject->addConfig('myApp', $configObject);
	}
);
```

And on the frontend this is available via that:

```js
const myAppsConfig = OCP.Config.getAppConfig('myApp');
```

Or did I miss something? Is this how you imagined it?

**Comment by ChristophWurst** [comment]
I wouldn't do it via an event but just a service like `IDataService` with a simple `addData(string $key, JsonSerializable $data)` method. That can be injected in app controllers -> apps loading data for their own pages, or more generic usage for apps that are loaded on all pages -> used in the `Application` initialization code or similar.

The client side would look like that, yes. Though I wouldn't call it "config" but more generic "data".

**Comment by MorrisJobke** [comment]
> loaded on all pages -> used in the Application initialization code or similar.

I don’t really like that because this means that it is also executed on all WebDAV requests 🙈

**Comment by ChristophWurst** [comment]
Fair enough. But that is still doable with such a service and a callback as parameter. Should the request result in a rendered template we invoke all passed callbacks.

In general I just prefer a properly typed interface (service) over the untyped and hard to discover event system.

**Comment by MorrisJobke** [comment]
@rullzer @ChristophWurst Mind to add this to the dev manual?

**Comment by rullzer** [comment]
@ChristophWurst is already busy with an extended fix. I have docs on my todo. This evening probably :+1:

**Comment by rullzer** [comment]
Done in https://github.com/nextcloud/documentation/pull/1209
